[receiver/k8scluster] Convert clusterresourcequota to use pdata

crobert-1 opened the pull request:
**Description:** <Describe what has changed.>

This converts the clusterresourcequota metrics of the k8scluster receiver to use the pdata metric format instead of opencensus.

**Link to tracking Issue:** <Issue number if applicable>
#4367

**Testing:** <Describe what testing was performed and which tests were added.>
Updated existing test to use pdata, ensuring metrics have same output format.

**Documentation:** <Describe the documentation added.>

crobert-1 committed:
[receiver/k8scluster] Convert clusterresourcequota to use pdata

This converts the clusterresourcequota metrics to use the pdata
format instead of opencensus.

crobert-1 commented on receiver/k8sclusterreceiver/internal/clusterresourcequota/clusterresourcequotas.go:
```diff
@@ -0,0 +1,62 @@
+// Copyright The OpenTelemetry Authors
```
I didn't actually delete `clusteresourcequoats.go`, and this isn't a whole new file, it's just a rename to fix the misspelling.

crobert-1 committed:
Run make generate

dmitryax commented on receiver/k8sclusterreceiver/internal/clusterresourcequota/clusteresourcequotas.go:
```diff
@@ -1,114 +0,0 @@
-		},
-		{
-			Key: conventions.AttributeK8SNamespaceName,
-		},
```
We need to preserve the existing format and not introduce breaking changes. The metrics are emitted with `resource` and `k8s.namespace.name` attributes. This PR drops the `resource` attribute and moves  `k8s.namespace.name` to a resource attribute.

crobert-1 replied:
Thanks, I've updated to match old format. However, I've run into a different issue when updating the existing test and I'm not sure what the best path forward here is.

Previously, multiple metrics of the same name would be emitted even though they share the same resource and scope, in the case of the `openshift.appliedclusterquota.limit` and `openshift.appliedclusterquota.used`. These metrics have an extra attribute for the `k8s.namespace.name` that the quota is for, but this is an attribute of the metric, not scope/resource. From [documentation](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#opentelemetry-protocol-data-model-producer-recommendations), only one metric of a given name should be emitted for a resource and scope, so the metric would just have multiple datapoints. After updating to pdata, only one metric will be emitted for a given name, and then each datapoint will have the given `k8s.namespace.name` as an attribute.

My question here is what's the best way to handle this? Do we try to match the old output even though it's wrong? Do we emit metrics properly and mark it a breaking change (if it is a breaking change)?

crobert-1 replied:
Also, I believe the `resourcequota` metrics had the same issue, but the unit test didn't cover this case. We shipped it with the [updated metric output](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23419), so I assume it's either okay to update to the proper format, or we'll need to update that.

dmitryax replied:
> Previously, multiple metrics of the same name would be emitted even though they share the same resource and scope, in the case of the openshift.appliedclusterquota.limit and openshift.appliedclusterquota.used.

Good finding. This is probably a bug of OC->OTel translation. Worth filing a ticket. We definitely don't need to recreate this wrong behavior. And mdatagen generated code won't even allow that. I wouldn't worry about applying it without marking it as a breaking change. It's a different representation (several metrics vs one metric with several datapoints), but the data is pretty much the same.

crobert-1 replied:
What should the bug be filed against? I'm not sure what the actual request would be if we're saying the old way was wrong, and now we're converting to the proper way? Is the action item to go through metrics that are still in the OpenCensus format and update them so they don't create multiple metrics with the same name, resource, and scope attributes?

dmitryax replied:
Actually, I misunderstood the root cause. It's not a bug in the OC->Otel conversion but rather a bug in how we have created the OC metrics in this receiver. So we are good now. Thanks!

dmitryax commented on receiver/k8sclusterreceiver/internal/clusterresourcequota/metadata.yaml:
```diff
@@ -0,0 +1,52 @@
+    description: The configured upper limit for a particular resource.
+    unit: "1"
+    gauge:
+      value_type: int
```
Please add the following attributes to maintain the current metrics format and regenerate:
- `resource` attribute to `openshift.clusterquota.limit` and `openshift.clusterquota.used` metrics
- `resource` and `k8s.namespace.name` attributes to `openshift.appliedclusterquota.limit` and `openshift.appliedclusterquota.used` metrics
Then you don't need to call `EmitForResource`

crobert-1 replied:
Thanks for pointing this out! I'll update. I'm planning on testing on an openshift cluster as well before marking this PR ready for to confirm the format hasn't changed. I'll also check to see if the e2e test for this receiver can be updated to test this, but since it's Openshift-specific I'm a bit skeptical.

dmitryax replied:
We cannot validate it with the e2e tests. Just use unit tests

crobert-1 commented on receiver/k8sclusterreceiver/internal/clusterresourcequota/clusterresourcequotas.go:
```diff
@@ -0,0 +1,62 @@
+	return mbphase.Emit()
+}
+
+func emitCRQMetrics(mbphase *imetadataphase.MetricsBuilder, crq *quotav1.ClusterResourceQuota, rl v1.ResourceList, metricName string, ts pcommon.Timestamp, namespace string) {
```
That's fair. I figured it would be nice to avoid 4 almost identical for-loops with the method, but I see your point.

dmitryax replied:
I don't think you need 4 loops. it should be only one

crobert-1 replied:
The loop is over the each of the `ClusterResourceQuota.ResourceList` objects. So we have to do the for loop over the resources governed by each of the given quota metrics.

`crq.Status.Total.Hard`, `crq.Status.Total.Used`, and then for each namespace within the cluster, do it namespace-by-namespace for `ns.Status.Hard` and `ns.Status.Used`.

dmitryax replied:
Just two nested loops don't work?

crobert-1 replied:
No, because the resources are under the specific quota "unit" in the schema here. So the type hierarchy is:
The `ClusterResourceQuota.Status.Hard` property has a list of resources where the value is the hard limits for the given. The `ClusterResourceQuota.Status.Used` property has a list of resources that shows `current observed total usage of the resource`.

The for loop is then:
```
foreach resource governed by hard limit:
  emit hard limit quota metrics with this resource
```

If the `ClusterResourceQuota` [type](https://pkg.go.dev/github.com/openshift/api@v0.0.0-20230720094506-afcbe27aec7c/quota/v1#ClusterResourceQuota) had a top level resource list that then showed the values for different quotas on that resource, we could have a single loop, or something like nested loops.

crobert-1 replied:
I've removed the extra method. After being able to use a single emit call, it doesn't looks as bad. Let me know if you'd prefer going back to the extra method though if this looks redundant.

crobert-1 committed:
Make changes requested

- resource should be an attribute of the metric, not a resource attribute.
- k8s namespace needs to also be an attribute of the metric and not
a resource attribute
- Remove unecessary extra method
- Update test to match original format of metrics

crobert-1 committed:
Fix linting failure

dmitryax commented on receiver/k8sclusterreceiver/internal/clusterresourcequota/clusterresourcequotas.go:
```diff
@@ -0,0 +1,61 @@
+		}
+
+		for k, v := range ns.Status.Used {
+			val := v.Value()
+			if strings.HasSuffix(string(k), ".cpu") {
+				val = v.MilliValue()
+			}
```
It's cleaner now. You can extract this part to a function `extractValue(k, v)`

crobert-1 replied:
Done.

crobert-1 committed:
Move re-used functionality into separate method

crobert-1 committed:
Fix linter

dmitryax commented on .chloggen/k8s-cluster-receiver-clusterresourcequota.yaml:
```diff
@@ -0,0 +1,21 @@
+note: "Change k8s.clusterresourcequota metrics to use mdatagen"
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [10553, 4367]
```
Not sure if we are contributing to https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/10553 in this PR

crobert-1 replied:
Good point, I removed it from the changelog and PR description.

dmitryax reviewed (approved):
LGTM

crobert-1 committed:
Fix import organization

dmitryax commented:
@crobert-1 is this PR ready?

crobert-1 committed:
Remove issue #10553 from CHANGELOG

crobert-1 commented:
> @crobert-1 is this PR ready?

Yes, this is ready.
